staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher.go

```diff
@@ -0,0 +1,268 @@
+func NewSchemaCacher(handler http.Handler, path string) *SchemaCacher {
+	return &SchemaCacher{
+		downloader: &handlerDownloader{
+			handler: handler,
+			path:    path,
+		},
+		parser: &schemaParser{},
+	}
+}
+
+// Get implements Getter
+func (s *SchemaCacher) Get() (Resources, error) {
+	err := s.updateSchema()
+	if err != nil {
+		return nil, err
+	}
+
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.openAPISchema, nil
+}
+
+// updateSchema parses the OpenAPI schema if it has changed since the last
+// time updateSchema was called, and cache it.
+func (s *SchemaCacher) updateSchema() error {
+	lastEtag := s.getLastEtag()
+	schemaBytes, newEtag, httpStatus, err := s.downloader.DownloadSchema(lastEtag)
+	if err != nil {
+		return err
+	}
+
+	// The schema hasn't changed so there's nothing to update
+	if httpStatus == http.StatusNotModified {
+		return nil
+	}
+
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+
+	// Check etag again after taking the write lock in case lastEtag
+	// was updated while waiting for the write lock
+	if newEtag == s.lastEtag {
+		return nil
+	}
+
+	// Convert raw spec to our Resource type
+	schema, err := s.parser.Parse(schemaBytes)
+	if err != nil {
+		return err
+	}
+
+	s.openAPISchema = schema
+	s.lastEtag = newEtag
+	return nil
+}
+
+func (s *SchemaCacher) getLastEtag() string {
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.lastEtag
+}
+
+// Downloader sepcifies where to download an OpenAPI schema from
+type Downloader interface {
+	// DownloadSchema attempts to get the OpenAPI schema if it has changed.
+	DownloadSchema(lastEtag string) (
+		schemaBytes []byte,
+		newEtag string,
+		httpStatus int,
+		err error)
+}
+
+type handlerDownloader struct {
+	handler http.Handler
+	path    string
+}
+
+var _ Downloader = &handlerDownloader{}
+
+// DownloadSchema implements Downloader
+func (d *handlerDownloader) DownloadSchema(lastEtag string) (schemaBytes []byte, newEtag string, httpStatus int, err error) {
+	req, err := http.NewRequest("GET", d.path, nil)
+	if err != nil {
+		return nil, "", 0, err
+	}
+	req.Header.Add("Accept", "application/com.github.proto-openapi.spec.v2@v1.0+protobuf")
+	if len(lastEtag) > 0 {
+		req.Header.Add("If-None-Match", lastEtag)
+	}
+
+	writer := utilresponsewriter.NewInMemoryResponseWriter()
+	d.handler.ServeHTTP(writer, req)
+
+	switch writer.RespCode {
+	case http.StatusNotModified:
+		if len(lastEtag) == 0 {
+			return nil, lastEtag, http.StatusNotModified, fmt.Errorf("http.StatusNotModified is not allowed in absence of etag")
+		}
+		return nil, lastEtag, http.StatusNotModified, nil
+	case http.StatusOK:
+		schemaBytes = writer.Data
+		newEtag = writer.Header().Get("Etag")
+		return schemaBytes, newEtag, http.StatusOK, nil
+	default:
+		return nil, "", 0, fmt.Errorf("failed to retrieve openAPI spec, http error: %s", writer.String())
+	}
+}
+
+// This interface is to here to make unit testing easier
+type Parser interface {
```

Comment:
The separation into multiple types is great, btw. Consider putting them in separate files.